src/crypto/sha256_avx512.cpp

```diff
@@ -0,0 +1,350 @@
+#ifdef ENABLE_AVX512
+
+#include <stdint.h>
+#include <immintrin.h>
+
+#include <crypto/sha256.h>
+#include <crypto/common.h>
+
+namespace sha256d64_avx512 {
+namespace {
+
+__m512i inline K(uint32_t x) { return _mm512_set1_epi32(x); }
+
+__m512i inline Add(__m512i x, __m512i y) { return _mm512_add_epi32(x, y); }
+__m512i inline Add(__m512i x, __m512i y, __m512i z) { return Add(Add(x, y), z); }
+__m512i inline Add(__m512i x, __m512i y, __m512i z, __m512i w) { return Add(Add(x, y), Add(z, w)); }
+__m512i inline Add(__m512i x, __m512i y, __m512i z, __m512i w, __m512i v) { return Add(Add(x, y, z), Add(w, v)); }
+__m512i inline Inc(__m512i& x, __m512i y) { x = Add(x, y); return x; }
+__m512i inline Inc(__m512i& x, __m512i y, __m512i z) { x = Add(x, y, z); return x; }
+__m512i inline Inc(__m512i& x, __m512i y, __m512i z, __m512i w) { x = Add(x, y, z, w); return x; }
+__m512i inline Xor(__m512i x, __m512i y) { return _mm512_xor_si512(x, y); }
+__m512i inline Xor(__m512i x, __m512i y, __m512i z) { return Xor(Xor(x, y), z); }
+__m512i inline Or(__m512i x, __m512i y) { return _mm512_or_si512(x, y); }
+__m512i inline And(__m512i x, __m512i y) { return _mm512_and_si512(x, y); }
+__m512i inline ShR(__m512i x, unsigned int n) { return _mm512_srli_epi32(x, n); }
```

Comment:
There is an implicit conversion from `unsigned int` to `int` here. Could be made explicit to make it easier to reason about correctness?

Reply:
gcc and Intel: __m512i _mm512_srli_epi32 (__m512i a, unsigned int imm8)
https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=_mm512_srli_epi32&expand=5515
https://github.com/gcc-mirror/gcc/blob/master/gcc/config/i386/avx512fintrin.h
clang: __m512i _mm512_srli_epi32(__m512i __A, int __B)
https://clang.llvm.org/doxygen/avx512fintrin_8h_source.html#l05168

defined to macro?
#define ShR _mm512_srli_epi32

Reply:
Oh, got it. Thanks for clarifying.

Also saw this old comment of mine: https://github.com/bitcoin/bitcoin/pull/13989#discussion_r221537873

Sorry for the confusion :-)